test/test_bonding.cpp

```diff
@@ -0,0 +1,149 @@
+{
+    const SRTSOCKET server_sock = srt_create_socket();
+    sockaddr_in bind_sa;
+    memset(&bind_sa, 0, sizeof bind_sa);
+    bind_sa.sin_family = AF_INET;
+    ASSERT_EQ(inet_pton(AF_INET, "127.0.0.1", &bind_sa.sin_addr), 1);
+    bind_sa.sin_port = htons(4200);
+
+    ASSERT_NE(srt_bind(server_sock, (sockaddr*)&bind_sa, sizeof bind_sa), -1);
+    const int yes = 1;
+    srt_setsockflag(server_sock, SRTO_GROUPCONNECT, &yes, sizeof yes);
+    ASSERT_NE(srt_listen(server_sock, 5), -1);
+    
+    std::this_thread::sleep_for(std::chrono::seconds(7));
+    // srt_accept..
+}
+
+TEST(Bonding, NonBlockingGroupConnect)
+{
+    srt_startup();
+    
+    const int ss = srt_create_group(SRT_GTYPE_BROADCAST);
+    ASSERT_NE(ss, SRT_ERROR);
+    std::cout << "Created group socket: " << ss << '\n';
+
+    int no = 0;
+    ASSERT_NE(srt_setsockopt(ss, 0, SRTO_RCVSYN, &no, sizeof no), SRT_ERROR); // non-blocking mode
+    ASSERT_NE(srt_setsockopt(ss, 0, SRTO_SNDSYN, &no, sizeof no), SRT_ERROR); // non-blocking mode
+
+    const int poll_id = srt_epoll_create();
+    // Will use this epoll to wait for srt_accept(...)
+    const int epoll_out = SRT_EPOLL_OUT | SRT_EPOLL_ERR;
+    ASSERT_NE(srt_epoll_add_usock(poll_id, ss, &epoll_out), SRT_ERROR);
+
+    // srt_connect_callback(m_sock, &TransmitGroupSocketConnect, this);
+
+    sockaddr_in sa;
+    sa.sin_family = AF_INET;
+    sa.sin_port = htons(4200);
+    ASSERT_EQ(inet_pton(AF_INET, "127.0.0.1", &sa.sin_addr), 1);
+
+    std::future<void> listen_promise = std::async(std::launch::async, listening_thread);
+    
+    std::cout << "Connecting two sockets " << std::endl;
+    for (int i = 0; i < 2; ++i)
+    {
+        const int sockid = srt_connect(ss, (sockaddr*) &sa, sizeof sa);
+        EXPECT_GT(sockid, 0) << "Socket " << i;
+        sa.sin_port = htons(4201); // Changing port so that second connect fails
+        std::cout << "Socket created: " << sockid << '\n';
+        ASSERT_NE(srt_epoll_add_usock(poll_id, sockid, &epoll_out), SRT_ERROR);
+    }
+    std::cout << "Returned from connecting two sockets " << std::endl;
+
+    const int default_len = 3;
+    int rlen = default_len;
+    SRTSOCKET read[default_len];
+
+    int wlen = default_len;
+    SRTSOCKET write[default_len];
+
+    for (int j = 0; j < 2; ++j)
+    {
+        const int epoll_res = srt_epoll_wait(poll_id, read, &rlen,
+            write, &wlen,
+            5000, /* timeout */
+            0, 0, 0, 0);
+            
+        std::cout << "Epoll result: " << epoll_res << '\n';
+        std::cout << "Epoll rlen: " << rlen << ", wlen: " << wlen << '\n';
+        for (int i = 0; i < rlen; ++i)
+        {
+            std::cout << "Epoll read[" << i << "]: " << read[i] << '\n';
+        }
+        for (int i = 0; i < wlen; ++i)
+        {
+            std::cout << "Epoll write[" << i << "]: " << write[i] << " (removed from epoll)\n";
+            EXPECT_EQ(srt_epoll_remove_usock(poll_id, write[i]), 0);
+        }
+    }
```

Comment:
In this test the second socket `261596698` connection should fail.

In the first iteration `epoll_wait(..)` notifies that two sockets are write-ready (connected) - GOOD.

On the second iteration, epoll has only one socket, that is expected to fail to connect. `epoll_wait(..)` returns 2, but the socket is not present in the `write[]` array.


`epoll_wait(..)` returns 2 does no returns:
```shell
[ RUN      ] Bonding.NonBlockingGroupConnect
Created group socket: 1335338525
Connecting two sockets 
Socket created: 261596700
Socket created: 261596698
Returned from connecting two sockets 
Epoll result: 2
Epoll rlen: 0, wlen: 2
Epoll write[0]: 261596700 (removed from epoll)
Epoll write[1]: 1335338525 (removed from epoll)

Epoll result: 2
Epoll rlen: 0, wlen: 1
Epoll write[0]: 261596698 (removed from epoll)
```